src/github/setup.ts

```diff
@@ -0,0 +1,115 @@
+import { createAppAuth } from "@octokit/auth-app";
+import { OctokitOptions } from "@octokit/core/dist-types/types";
+import { ThrottlingOptions } from "@octokit/plugin-throttling/dist-types/types";
+import { Octokit } from "@octokit/rest";
+import { Logger, validate } from "opstooling-js";
+
+import { GitHubConfigOpts } from "src/github";
+import {
+  GitHubAppAuthSchema,
+  GitHubAppInstallationAuthSchema,
+  GitHubTokenAuthSchema,
+} from "src/schemas/GitHubConfigEnvSchema";
+import { GitHubAppAuthEnv, GitHubAppInstallationAuthEnv, GitHubTokenAuthEnv } from "src/types/generated";
+import { lazyApi } from "src/utils";
+
+export function getInstance(opts: GitHubConfigOpts): Promise<Octokit> {
+  let authStrategy: OctokitOptions["authStrategy"];
+  let auth: OctokitOptions["auth"];
+
+  switch (opts.authType) {
+    case "app":
+    case "installation":
+      authStrategy = createAppAuth;
+      auth = {
+        appId: opts.appId,
+        clientId: opts.clientId,
+        clientSecret: opts.clientSecret,
+        privateKey: opts.privateKey,
+      };
+      break;
+    case "token":
+      auth = opts.authToken;
+  }
+
+  const throttle: ThrottlingOptions = {
+    // Options are typed as `any` in octokit/plugin-throttling
+    /* eslint-disable @typescript-eslint/no-explicit-any,@typescript-eslint/restrict-template-expressions */
+    onRateLimit: (retryAfter, options: any, octokit, retryCount) => {
+      octokit.log.warn(`Request quota exhausted for request ${options.method} ${options.url}`);
+
+      if (retryCount < 3) {
+        octokit.log.info(`Retrying after ${retryAfter} seconds!`);
+        return true;
+      }
+    },
+    onSecondaryRateLimit: (retryAfter, options: any, octokit, retryCount) => {
+      // does not retry, only logs a warning
+      octokit.log.warn(`SecondaryRateLimit detected for request ${options.method} ${options.url}`);
+
+      if (retryCount < 3) {
```

Comment:
nit: might be good to have it configurable from opts?

Reply:
Let's do that when first needed.